import {
  ArgumentsHost,
  Catch,
  ExceptionFilter,
  HttpException,
  HttpStatus,
  BadRequestException,
} from '@nestjs/common'
import { Response } from 'express'
import { AppException } from '../errors/app.exception'
import { ValidationException } from 'src/exceptions/validation.exception'

@Catch()
export class GlobalExceptionFilter implements ExceptionFilter {
  catch(exception: any, host: ArgumentsHost) {
    const ctx = host.switchToHttp()
    const response = ctx.getResponse<Response>()

    let status = HttpStatus.INTERNAL_SERVER_ERROR
    let errorResponse = {
      code: 'INTERNAL_SERVER_ERROR',
      message: '–í–Ω—É—Ç—Ä–µ–Ω–Ω—è—è –æ—à–∏–±–∫–∞ —Å–µ—Ä–≤–µ—Ä–∞',
    }

    if (exception instanceof AppException) {
      status = HttpStatus.INTERNAL_SERVER_ERROR
      errorResponse = {
        code: exception.code,
        message: exception.message,
      }
    }

    // üëâ –û–±—Ä–∞–±–æ—Ç–∫–∞ –≤–∞–ª–∏–¥–∞—Ü–∏–æ–Ω–Ω—ã—Ö –æ—à–∏–±–æ–∫ –æ—Ç class-validator
    else if (exception instanceof BadRequestException) {
      const res = exception.getResponse()
      const messages = typeof res === 'object' && res['message'] ? res['message'] : ['–û—à–∏–±–∫–∞ –≤–∞–ª–∏–¥–∞—Ü–∏–∏']

      status = HttpStatus.BAD_REQUEST
      errorResponse = {
        code: 'VALIDATION_ERROR',
        message: Array.isArray(messages) ? messages.join(', ') : messages,
      }
    } else if (exception instanceof ValidationException) {
      status = HttpStatus.BAD_REQUEST
      errorResponse = {
        code: 'VALIDATION_ERROR',
        message: exception.message,
      }
    } else if (exception instanceof HttpException) {
      status = exception.getStatus()
      const res: any = exception.getResponse()
      errorResponse = {
        code: String(status),
        message: typeof res === 'string' ? res : res.message || '–û—à–∏–±–∫–∞',
      }
    }

    response.status(status).json({
      success: false,
      data: undefined,
      error: errorResponse,
    })
  }
}
